### PR TITLE
split_array: remove runtime panics

### PR DESCRIPTION
**Pull request by Xiretza**
This is an update to #90091 implementing https://github.com/rust-lang/rust/issues/90091#issuecomment-1130349259, which removes runtime panics by:
- making slice-splitting methods return an `Option`
- asserting the index bounds of array-splitting methods using a `const` assert.

The latter is a temporary measure needed because generic const expressions are not yet advanced enough to represent this restriction at the type level. As such it is not supposed to be stabilized as-is. It is however an improvement over the previous behaviour, which was an unconditional runtime panic instead of a compile-time error.

### CODE REVIEW

**Comment by Xiretza** [comment]
Manged to make everything except `array::split_array()` `const`, maybe there's a better way to implement that one that doesn't rely on `ManuallyDrop::deref()`?

**Comment by BoxyUwU** [comment]
`generic_const_exprs` isn't allowed to be used in `core`/`std` as its too unstable/broken

**Comment by Xiretza** [comment]
Alright, at least we now have an answer to that question now.

File: library/core/src/array/mod.rs [hunk 1]
```diff
@@ -777,17 +821,70 @@ impl<T, const N: usize> [T; N] {
...
+        // - `right` is a slice of `N - (N-M) == M` elements.
+        unsafe { (from_mut_slice_unchecked(left), from_mut_slice_unchecked(right)) }
     }
 }
 
+/// Unsafely converts a slice of `N` elements to an array reference of `N` elements.
+///
+/// # Safety
+///
+/// The caller must ensure that the slice has exactly `N` elements. Violating this constraint
```
Comment (Kixunil) [inline library/core/src/array/mod.rs:841]: I think this should say "the slice has at least `N` elements". Calling it is sound if it has more.

File: library/core/src/array/mod.rs [hunk 2]
```diff
@@ -777,17 +821,70 @@ impl<T, const N: usize> [T; N] {
...
+const unsafe fn from_slice_unchecked<T, const N: usize>(s: &[T]) -> &[T; N] {
+    // SAFETY: caller guarantees that `s` is a slice of `N` elements.
+    unsafe { &*(s.as_ptr() as *const [T; N]) }
+}
+
+/// Unsafely converts a mutable slice of `N` elements to a mutable array reference of `N` elements.
+///
+/// # Safety
+///
+/// The caller must ensure that the slice has exactly `N` elements. Violating this constraint
```
Comment (Kixunil) [inline library/core/src/array/mod.rs:865]: Same as above.

**Comment by Kixunil** [comment]
> maybe there's a better way to implement that one that doesn't rely on `ManuallyDrop::deref()`?

I think you can just call `mem::forget` on the original array after copying. Since the reads don't panic this should be sound.

File: library/core/src/slice/mod.rs [hunk 3]
```diff
@@ -1797,10 +1795,15 @@ impl<T> [T] {
     #[inline]
     #[track_caller]
     #[must_use]
-    pub fn split_array_ref<const N: usize>(&self) -> (&[T; N], &[T]) {
-        let (a, b) = self.split_at(N);
-        // SAFETY: a points to [T; N]? Yes it's [T] of length N (checked by split_at)
-        unsafe { (&*(a.as_ptr() as *const [T; N]), b) }
+    pub const fn split_array_ref<const N: usize>(&self) -> Option<(&[T; N], &[T])> {
```
[the remark this answers is not in the document]
Comment (Xiretza) [reply library/core/src/slice/mod.rs:1798]: That would blow up the method count to 8; I don't think adding `unwrap()`/`except()` is that big of a hurdle.

**Comment by Xiretza** [comment]
> I think you can just call `mem::forget` on the original array after copying

Thanks, it's `const` now!

**Comment by Xiretza** [comment]
I've trimmed this down to remove the need for `#![feature(generic_const_exprs)]`, the array methods also return `Option<(&[T; M], &[T])>` instead of `(&[T; M], &[T; N-M])` for now. This at least removes the panicking, making invalid split indices on arrays a compile-time error is future work.

@rustbot ready

**Comment by faern** [comment]
> the array methods also return Option<(&[T; M], &[T])> instead of (&[T; M], &[T; N-M]) for now. This at least removes the panicking, making invalid split indices on arrays a compile-time error is future work.

This makes the function signature wrong. I'd argue that's worse than the panic. And you don't remove the panic, you just move the responsibility of doing that to the caller. I'd argue that it would be better if the function signature was at least correct now, even if the exact desired behavior can't be had yet.

**Comment by Xiretza** [comment]
> This makes the function signature wrong.

The function signature for the array methods were already wrong, now they're just wrong in a slightly different way. They cannot be made correct yet, that's what `#![feature(generic_const_exprs)]` was for.

> And you don't remove the panic, you just move the responsibility of doing that to the caller.

I guess the array methods could use a `static_assert!()` and impossible-to-fail internal `.unwrap()` instead to approximate the intended final behaviour, I'll look into that.

**Comment by Xiretza** [comment]
Updated to @faern's suggestions - the array method signatures are not changed anymore now, but the implementations contain a const assert that at least turns index errors from runtime panics into post-mono errors.

**Comment by workingjubilee** [comment]
Can you fully change the message on the title and PR to describe what's actually the case right now? In this case, technically these functions are in fact "panicking", but now at compile-time. The bors merge commit will include that message, so including a bunch of Markdown-annotated strikethrough is not useful.

**Comment by workingjubilee** [comment]
I am not really aware of any other code in the standard library that deliberately relies on user-facing post-monomorphization errors. I believe this may be the first? I believe this should be discussed by the relevant teams before accepting this PR.

**Comment by Xiretza** [comment]
Updated the title and description. Note that the design of the array methods as implemented here is not intended to be final, only an improvement over the current implementation.

**Review by dtolnay — COMMENTED** [review]
We discussed this PR in today's library API meeting. Thank you for the PR!

There are 2 changes here:

- Methods that changed from runtime panic on out of bounds, to const assert. We are not on board with this. This is in the same category as https://github.com/rust-lang/rust/pull/99471#issuecomment-1555908856.

    Instead we'd recommend that this be integrated into the existing `unconditional_panic` lint which applies to things like `array[100]` out of bounds.

- Methods that changed from runtime panic to Option return type. We are also not on board with this. We preferred that the `split_array`-related APIs be viewed similarly to `str::split_at` (https://doc.rust-lang.org/1.71.0/std/primitive.str.html#method.split_at), or `slice::split_at`, which panic.

    Yes, making the user transform an Option into a panic is easier for them than the reverse, but we expect that the dominant use cases will be such that the user _knows_ this index is in bounds, and proliferation of `unwrap()` in such code is unpleasant.

    When the index is not known to be in bounds at the call site, going through the iteration API to pull off a fixed sized chunk (https://github.com/rust-lang/rust/issues/74985) seems like a suitable fit, and the documentation could guide this.

**Comment by Xiretza** [comment]
@dtolnay This is incredibly frustrating.

I [proposed this](https://github.com/rust-lang/rust/issues/90091#issuecomment-1130349259) in the tracking issue over a year ago. Judging by the fact that it has over half as many :+1: reactions as the tracking issue itself, it seemed to be rather well received by the community, i.e. the actual prospective users of the API.

People kept bringing up that the methods should be fallible, so after almost a year of nobody with an actual say in things showing up to say whether it would be acceptable, I finally opened this PR half a year ago. Got some initial reviews that answered a question that was also asked [on the tracking issue](https://github.com/rust-lang/rust/issues/90091#issuecomment-1287674439) half a year prior without any real answers there, but at least then I knew and could fix my implementation.

There were crickets for another couple months, when out of the blue libs-api shows up to shoot it down entirely. Meanwhile there has still not been a single official comment on the tracking issue - am I just misunderstanding what tracking issues are for?

This is frustrating not only because of the time I wasted, but also because it feels like I have a need for the fallible API just about every month, but have not once wished for the panicking API. `array_chunks` is also not an alternative, as it produces chunks that are all the same size.

**Comment by faern** [comment]
> We preferred that the split_array-related APIs be viewed similarly to str::split_at (https://doc.rust-lang.org/1.71.0/std/primitive.str.html#method.split_at), or slice::split_at, which panic.

I think this is a mistake. I personally really don't like that `split_at` panics. Every time I reach for that method I wish it was fallible

> but we expect that the dominant use cases will be such that the user knows this index is in bounds, and proliferation of unwrap() in such code is unpleasant.

I definitely don't. Because the split index can be program input that I don't control at compile time. I can't remember a single time where I wanted it to be panicking. Working with byte buffers in memory has so many papercuts in Rust still, and good fallible slice/array subdivision in various kinds is a major part of those.

**Comment by Kixunil** [comment]
Exactly, `split_array_ref` has a special use case in parsers where we want to get a known length array. Example:

```rust
let (value, input) = input.split_arrray_ref::<4>().ok_or(Error::UnexpectedEof)?;
let value = u32::from_be_byte(value);
```

With panic this turns into DoS vulnerability hazard.

**Pull request closed without merging**